src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyName.cs

```diff
@@ -0,0 +1,393 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+using System.Numerics;
+using System.Reflection;
+
+namespace System.Runtime.Loader
+{
+    internal unsafe struct AssemblyMetaDataInternal
+    {
+        public ushort usMajorVersion;         // Major Version.
+        public ushort usMinorVersion;         // Minor Version.
+        public ushort usBuildNumber;          // Build Number.
+        public ushort usRevisionNumber;       // Revision Number.
+        public byte* szLocale;               // Locale.
+    }
+
+    internal enum CorAssemblyFlags
+    {
+        afPublicKey = 0x0001,     // The assembly ref holds the full (unhashed) public key.
+
+        afPA_None = 0x0000,     // Processor Architecture unspecified
+        afPA_MSIL = 0x0010,     // Processor Architecture: neutral (PE32)
+        afPA_x86 = 0x0020,     // Processor Architecture: x86 (PE32)
+        afPA_IA64 = 0x0030,     // Processor Architecture: Itanium (PE32+)
+        afPA_AMD64 = 0x0040,     // Processor Architecture: AMD X64 (PE32+)
+        afPA_ARM = 0x0050,     // Processor Architecture: ARM (PE32)
+        afPA_ARM64 = 0x0060,     // Processor Architecture: ARM64 (PE32+)
+        afPA_NoPlatform = 0x0070,      // applies to any platform but cannot run on any (e.g. reference assembly), should not have "specified" set
+        afPA_Specified = 0x0080,     // Propagate PA flags to AssemblyRef record
+        afPA_Mask = 0x0070,     // Bits describing the processor architecture
+        afPA_FullMask = 0x00F0,     // Bits describing the PA incl. Specified
+        afPA_Shift = 0x0004,     // NOT A FLAG, shift count in PA flags <--> index conversion
+
+        afEnableJITcompileTracking = 0x8000, // From "DebuggableAttribute".
+        afDisableJITcompileOptimizer = 0x4000, // From "DebuggableAttribute".
+        afDebuggableAttributeMask = 0xc000,
+
+        afRetargetable = 0x0100,     // The assembly can be retargeted (at runtime) to an
+                                     //  assembly from a different publisher.
+
+        afContentType_Default = 0x0000,
+        afContentType_WindowsRuntime = 0x0200,
+        afContentType_Mask = 0x0E00, // Bits describing ContentType
+    }
+
+    internal enum AssemblyNameIncludeFlags
+    {
+        INCLUDE_DEFAULT = 0x00,
+        INCLUDE_VERSION = 0x01,
+        INCLUDE_ARCHITECTURE = 0x02,
+        INCLUDE_RETARGETABLE = 0x04,
+        INCLUDE_CONTENT_TYPE = 0x08,
+        INCLUDE_PUBLIC_KEY_TOKEN = 0x10,
+        EXCLUDE_CULTURE = 0x20,
+        INCLUDE_ALL = INCLUDE_VERSION
+                    | INCLUDE_ARCHITECTURE
+                    | INCLUDE_RETARGETABLE
+                    | INCLUDE_CONTENT_TYPE
+                    | INCLUDE_PUBLIC_KEY_TOKEN,
+    }
+
+    internal sealed unsafe class BinderAssemblyName : AssemblyIdentity, IEquatable<BinderAssemblyName>
```

Comment:
This should be replaced with `System.Reflection.AssemblyName` in following PR. It may require more native side changes so I'm not willing to include in this PR.

The `BINDER_SPACE::AssemblyName` type is still used in native. The conversion should cover all its usage.

Reply:
An interesting thing is that `S.R.AssemblyName` can be modified by malicious event handler on ALC. I'm not sure about the scope it's affecting. Anyway, cloning shouldn't more complex than creating a new type.